src/simularium/VisDataCache.ts

```diff
@@ -0,0 +1,270 @@
+    }
+
+    private frameAccessError(msg?: string): CachedFrame {
+        throw new FrontEndError(
+            `Error accessing frame: ${msg}`,
+            ErrorLevel.WARNING
+        );
+    }
+
+    public hasFrames(): boolean {
+        return this.numFrames > 0 && this.head !== null && this.tail !== null;
+    }
+
+    /**
+     * Walks the cache looking for node that satisfies condition
+     * returns the node if found, otherwise returns null,
+     * starts at head if firstNode is not provided.
+     */
+    private walkLinkedList(
+        condition: (data: CacheNode) => boolean,
+        firstNode?: CacheNode
+    ): CacheNode | null {
+        let currentNode = firstNode || this.head;
+        while (currentNode) {
+            if (condition(currentNode)) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+        }
+        return null;
+    }
+
+    public containsTime(time: number): boolean {
+        if (
+            compareTimes(time, this.getFirstFrameTime(), 0.1) === -1 ||
+            compareTimes(time, this.getLastFrameTime(), 0.1) === 1
+        ) {
+            return false;
+        }
+        if (time < this.getFirstFrameTime() || time > this.getLastFrameTime()) {
+            return false;
+        }
+        return !!this.walkLinkedList((node) => node.data.time === time);
+    }
+
+    public containsFrameAtFrameNumber(frameNumber: number): boolean {
+        if (
+            frameNumber < this.getFirstFrameNumber() ||
+            frameNumber > this.getLastFrameNumber()
+        ) {
+            return false;
+        }
+        return !!this.walkLinkedList(
+            (node) => node.data.frameNumber === frameNumber
+        );
+    }
+
+    public getFirstFrame(): CachedFrame {
+        if (!this.head) {
+            return this.frameAccessError("No data in cache.");
+        }
+        return this.head.data;
+    }
+
+    public getFirstFrameNumber(): number {
+        return this.head?.data.frameNumber || -1;
+    }
+
+    public getFirstFrameTime(): number {
+        return this.head?.data.time || -1;
+    }
+
+    public getLastFrame(): CachedFrame {
+        if (!this.tail) {
+            return this.frameAccessError(" No data in cache.");
+        }
+        return this.tail.data;
+    }
+
+    public getLastFrameNumber(): number {
+        return this.tail?.data.frameNumber || -1;
+    }
+
+    public getLastFrameTime(): number {
+        return this.tail?.data.time || -1;
+    }
+
+    private getFrameAtCondition(
+        condition: (data: CacheNode) => boolean,
+        value: number
+    ): CachedFrame {
+        if (!this.head) {
+            return this.frameAccessError("No data in cache.");
+        }
+        const frame = this.walkLinkedList(condition);
+        if (frame) {
+            return frame.data;
+        }
+        return this.frameAccessError(
+            `Frame not found at provided ${condition}. Attempting to access frame ${value}.`
```

Comment:
yeah putting `condition` in the string does seem a bit extreme!

Reply:
Hm, I see what you mean, also this isn't actually correct given some other changes that were made, since we could be looking for a `time` and not a `frameNumber`